flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/metrics/KafkaSourceReaderMetricsListener.java

```diff
@@ -0,0 +1,174 @@
+
+    public KafkaSourceReaderMetricsListener() {
+        TestingMetricRegistry registry =
+                TestingMetricRegistry.builder()
+                        .setDelimiter(DELIMITER)
+                        .setRegisterConsumer(
+                                (metric, metricName, group) -> {
+                                    String parentGroupScope =
+                                            group.getLogicalScope(CharacterFilter.NO_OP_FILTER);
+
+                                    // We only care about Kafka source reader metrics
+                                    if (isKafkaSourceReaderMetric(parentGroupScope)) {
+
+                                        // KafkaConsumer metric
+                                        if (isKafkaConsumerMetric(parentGroupScope)) {
+                                            kafkaConsumerMetrics.put(metricName, metric);
+                                            return;
+                                        }
+
+                                        final String topic =
+                                                group.getAllVariables()
+                                                        .get(
+                                                                ScopeFormat.asVariable(
+                                                                        KafkaSourceReaderMetrics
+                                                                                .TOPIC_GROUP));
+                                        final String partition =
+                                                group.getAllVariables()
+                                                        .get(
+                                                                ScopeFormat.asVariable(
+                                                                        KafkaSourceReaderMetrics
+                                                                                .PARTITION_GROUP));
+
+                                        if (topic == null && partition == null) {
+                                            // General metric
+                                            kafkaSourceReaderMetrics.put(metricName, metric);
+                                        } else {
+                                            // topic partition specific metric
+                                            kafkaSourceReaderMetrics.put(
+                                                    String.join(
+                                                            String.valueOf(DELIMITER),
+                                                            topic,
+                                                            partition,
+                                                            metricName),
+                                                    metric);
+                                        }
+                                    }
+                                })
```

Comment:
It seems that the code length is too long to express with lambda. Wouldn't it be better to separate it into a function?

Reply:
Thanks for your review Myoungdo! Fixed in 4c65357.